### PR TITLE
feat: Add Dedup Tables

### PR DESCRIPTION
**Pull request by antonyip**
Solution to De-duplicating the chainwalker tables.

First run
```
root@cc79d4019fff:/harmony# dbt run --select core.txs_deduped
03:17:53  Running with dbt=1.0.0
03:17:53  Found 20 models, 16 tests, 0 snapshots, 2 analyses, 182 macros, 0 operations, 1 seed file, 2 sources, 0 exposures, 0 metrics
03:17:53  
03:18:04  Concurrency: 4 threads (target='dev')
03:18:04  
03:18:04  1 of 1 START incremental model DEV.txs_deduped.................................. [RUN]
03:42:40  1 of 1 OK created incremental model DEV.txs_deduped............................. [SUCCESS 1 in 1476.03s]
03:42:40  
03:42:40  Finished running 1 incremental model in 1486.48s.
03:42:40  
03:42:40  Completed successfully
03:42:40  
03:42:40  Done. PASS=1 WARN=0 ERROR=0 SKIP=0 TOTAL=1
```

Second Run
```
root@cc79d4019fff:/harmony# dbt run --select core.txs_deduped
03:46:13  Running with dbt=1.0.0
03:46:13  [WARNING]: Configuration paths exist in your dbt_project.yml file which do not apply to any resources.
There are 1 unused configuration paths:
- seeds
03:46:14  Found 20 models, 39 tests, 0 snapshots, 2 analyses, 182 macros, 0 operations, 0 seed files, 2 sources, 0 exposures, 0 metrics
03:46:14  
03:46:27  Concurrency: 4 threads (target='dev')
03:46:27  
03:46:27  1 of 1 START incremental model DEV.txs_deduped.................................. [RUN]
03:46:38  1 of 1 OK created incremental model DEV.txs_deduped............................. [SUCCESS 1 in 10.65s]
03:46:38  
03:46:38  Finished running 1 incremental model in 24.40s.
03:46:38  
03:46:38  Completed successfully
03:46:38  
03:46:38  Done. PASS=1 WARN=0 ERROR=0 SKIP=0 TOTAL=1
```

Tests
```
root@0d5ab8255afd:/harmony# dbt test
03:51:10  Running with dbt=1.0.0
03:51:10  [WARNING]: Configuration paths exist in your dbt_project.yml file which do not apply to any resources.
There are 1 unused configuration paths:
- seeds

03:51:10  Found 20 models, 36 tests, 0 snapshots, 2 analyses, 182 macros, 0 operations, 0 seed files, 2 sources, 0 exposures, 0 metrics
03:51:10  
03:51:15  Concurrency: 4 threads (target='dev')
03:51:15  
03:51:15  1 of 36 START test not_null_blocks_block_hash................................... [RUN]
03:51:15  2 of 36 START test not_null_blocks_block_id..................................... [RUN]
03:51:15  3 of 36 START test not_null_blocks_block_parent_hash............................ [RUN]
03:51:15  4 of 36 START test not_null_blocks_block_timestamp.............................. [RUN]
03:51:21  1 of 36 PASS not_null_blocks_block_hash......................................... [PASS in 5.79s]
03:51:21  5 of 36 START test not_null_blocks_deduped_block_id............................. [RUN]
03:51:21  3 of 36 PASS not_null_blocks_block_parent_hash.................................. [PASS in 5.83s]
03:51:21  6 of 36 START test not_null_blocks_deduped_block_timestamp...................... [RUN]
03:51:21  2 of 36 PASS not_null_blocks_block_id........................................... [PASS in 5.98s]
03:51:21  7 of 36 START test not_null_blocks_deduped_chain_id............................. [RUN]
03:51:22  4 of 36 PASS not_null_blocks_block_timestamp.................................... [PASS in 6.95s]
03:51:22  8 of 36 START test not_null_blocks_deduped_header............................... [RUN]
03:51:25  6 of 36 PASS not_null_blocks_deduped_block_timestamp............................ [PASS in 3.94s]
03:51:25  9 of 36 START test not_null_blocks_deduped_ingested_at.......................... [RUN]
03:51:25  7 of 36 PASS not_null_blocks_deduped_chain_id................................... [PASS in 3.94s]
03:51:25  10 of 36 START test not_null_blocks_deduped_network............................. [RUN]
03:51:27  5 of 36 PASS not_null_blocks_deduped_block_id................................... [PASS in 5.70s]
03:51:27  11 of 36 START test not_null_blocks_deduped_offset_id........................... [RUN]
03:51:29  10 of 36 PASS not_null_blocks_deduped_network................................... [PASS in 3.67s]
03:51:29  12 of 36 START test not_null_blocks_deduped_record_id........................... [RUN]
03:51:31  11 of 36 PASS not_null_blocks_deduped_offset_id................................. [PASS in 3.77s]
03:51:31  13 of 36 START test not_null_blocks_deduped_tx_count............................ [RUN]
03:51:31  9 of 36 PASS not_null_blocks_deduped_ingested_at................................ [PASS in 6.17s]
03:51:31  14 of 36 START test not_null_blocks_gas_limit................................... [RUN]
03:51:33  12 of 36 PASS not_null_blocks_deduped_record_id................................. [PASS in 3.68s]
03:51:33  15 of 36 START test not_null_blocks_gas_used.................................... [RUN]
03:51:34  13 of 36 PASS not_null_blocks_deduped_tx_count.................................. [PASS in 3.75s]
03:51:34  16 of 36 START test not_null_blocks_miner....................................... [RUN]
03:51:36  14 of 36 PASS not_null_blocks_gas_limit......................................... [PASS in 4.36s]
03:51:36  17 of 36 START test not_null_blocks_nonce....................................... [RUN]
03:51:36  8 of 36 PASS not_null_blocks_deduped_header..................................... [PASS in 14.02s]
03:51:36  18 of 36 START test not_null_blocks_receipts_root............................... [RUN]
03:51:40  17 of 36 PASS not_null_blocks_nonce............................................. [PASS in 4.36s]
03:51:40  19 of 36 START test not_null_blocks_size........................................ [RUN]
03:51:40  16 of 36 PASS not_null_blocks_miner............................................. [PASS in 5.96s]
03:51:40  20 of 36 START test not_null_blocks_state_root.................................. [RUN]
03:51:41  18 of 36 PASS not_null_blocks_receipts_root..................................... [PASS in 4.41s]
03:51:41  21 of 36 START test not_null_blocks_tx_count.................................... [RUN]
03:51:45  21 of 36 PASS not_null_blocks_tx_count.......................................... [PASS in 3.84s]
03:51:45  22 of 36 START test not_null_metric_miner_count_metric_date..................... [RUN]
03:51:45  15 of 36 PASS not_null_blocks_gas_used.......................................... [PASS in 12.56s]
03:51:45  23 of 36 START test not_null_metric_miner_count_metric_period................... [RUN]
03:51:47  19 of 36 PASS not_null_blocks_size.............................................. [PASS in 7.25s]
03:51:47  24 of 36 START test not_null_metric_miner_count_miner_count..................... [RUN]
03:51:49  22 of 36 PASS not_null_metric_miner_count_metric_date........................... [PASS in 4.83s]
03:51:49  25 of 36 START test not_null_txs_deduped_block_id............................... [RUN]
03:51:50  20 of 36 PASS not_null_blocks_state_root........................................ [PASS in 9.40s]
03:51:50  26 of 36 START test not_null_txs_deduped_block_timestamp........................ [RUN]
03:51:51  23 of 36 PASS not_null_metric_miner_count_metric_period......................... [PASS in 5.50s]
03:51:51  27 of 36 START test not_null_txs_deduped_chain_id............................... [RUN]
03:51:51  24 of 36 PASS not_null_metric_miner_count_miner_count........................... [PASS in 4.00s]
03:51:51  28 of 36 START test not_null_txs_deduped_ingested_at............................ [RUN]
03:51:54  26 of 36 PASS not_null_txs_deduped_block_timestamp.............................. [PASS in 3.98s]
03:51:54  29 of 36 START test not_null_txs_deduped_network................................ [RUN]
03:51:54  25 of 36 PASS not_null_txs_deduped_block_id..................................... [PASS in 4.60s]
03:51:54  30 of 36 START test not_null_txs_deduped_offset_id.............................. [RUN]
03:51:54  27 of 36 PASS not_null_txs_deduped_chain_id..................................... [PASS in 3.76s]
03:51:54  31 of 36 START test not_null_txs_deduped_record_id.............................. [RUN]
03:51:58  30 of 36 PASS not_null_txs_deduped_offset_id.................................... [PASS in 3.76s]
03:51:58  32 of 36 START test not_null_txs_deduped_tx_block_index......................... [RUN]
03:51:59  29 of 36 PASS not_null_txs_deduped_network...................................... [PASS in 5.22s]
03:51:59  33 of 36 START test not_null_txs_deduped_tx_id.................................. [RUN]
03:51:59  31 of 36 PASS not_null_txs_deduped_record_id.................................... [PASS in 4.57s]
03:51:59  34 of 36 START test unique_blocks_block_hash.................................... [RUN]
03:52:02  32 of 36 PASS not_null_txs_deduped_tx_block_index............................... [PASS in 3.91s]
03:52:02  35 of 36 START test unique_blocks_block_id...................................... [RUN]
03:52:04  28 of 36 PASS not_null_txs_deduped_ingested_at.................................. [PASS in 13.03s]
03:52:04  36 of 36 START test unique_blocks_block_parent_hash............................. [RUN]
03:52:05  33 of 36 PASS not_null_txs_deduped_tx_id........................................ [PASS in 5.83s]
03:52:05  34 of 36 PASS unique_blocks_block_hash.......................................... [PASS in 5.89s]
03:52:06  35 of 36 PASS unique_blocks_block_id............................................ [PASS in 4.00s]
03:52:09  36 of 36 PASS unique_blocks_block_parent_hash................................... [PASS in 4.30s]
03:52:09  
03:52:09  Finished running 36 tests in 58.40s.
03:52:09  
03:52:09  Completed successfully
03:52:09  
03:52:09  Done. PASS=36 WARN=0 ERROR=0 SKIP=0 TOTAL=36
```

### CODE REVIEW

File: models/core/blocks_deduped.sql [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+        unique_key='block_id',
+        tags=['core'],
+        cluster_by=['block_timestamp']
+        )
+}}
+
+with deduped_raw_blocks as (
+
+    select 
+    *
```
Comment (chuxinh) [inline models/core/blocks_deduped.sql]: indent

File: models/core/blocks_deduped.sql [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    from {{ source("chainwalkers", "harmony_blocks" ) }}
+    where {{ incremental_load_filter("block_timestamp") }}
+    qualify row_number() over (partition by block_id order by ingested_at desc) = 1
+
+)
+
+select * from deduped_raw_blocks
```
Comment (chuxinh) [inline models/core/blocks_deduped.sql]: add extra line below

File: models/core/blocks_deduped.yml [hunk 3]
```diff
@@ -0,0 +1,50 @@
+version: 2
+models: 
+  - name: blocks_deduped
+    description: |-
+      Blocks from the Harmony chain shard 0. Duplicates have been removed.
+    columns:
+      - name: record_id
+        description: A unique id for the record generated by Chainwalkers
```
Comment (chuxinh) [inline models/core/blocks_deduped.yml]:
```suggestion
        description: A unique id for the record generated by Chainwalkers.
```

File: models/core/blocks_deduped.yml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        tests:
+          - not_null
+
+      - name: offset_id
+        description: Synonmous with block_id for Harmony
```
Comment (chuxinh) [inline models/core/blocks_deduped.yml]:
```suggestion
        description: Synonmous with `block_id` for Harmony.
```

File: models/core/blocks_deduped.yml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        tests:
+          - not_null
+
+      - name: block_id
+        description: The height of the chain this block corresponds with
```
Comment (chuxinh) [inline models/core/blocks_deduped.yml]:
```suggestion
        description: The height of the chain this block corresponds with.
```

File: models/core/blocks_deduped.yml [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        tests:
+          - not_null
+
+      - name: block_timestamp
+        description: The time the block was minted
```
Comment (chuxinh) [inline models/core/blocks_deduped.yml]:
```suggestion
        description: The time the block was minted.
```

File: models/core/blocks_deduped.yml [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        tests:
+          - not_null
+
+      - name: network
+        description: The blockchain network (i.e. mainnet, testnet, etc.)
```
Comment (chuxinh) [inline models/core/blocks_deduped.yml]:
```suggestion
        description: The blockchain network (i.e. mainnet, testnet, etc.).
```

File: models/core/blocks_deduped.yml [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        tests:
+          - not_null
+
+      - name: chain_id
+        description: Synonmous with blockchain name for Harmony
```
Comment (chuxinh) [inline models/core/blocks_deduped.yml]:
```suggestion
        description: Synonmous with blockchain name for Harmony.
```

File: models/core/blocks_deduped.yml [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        tests:
+          - not_null
+
+      - name: tx_count
+        description: The number of transactions in the block
```
Comment (chuxinh) [inline models/core/blocks_deduped.yml]:
```suggestion
        description: The number of transactions in the block.
```

File: models/core/blocks_deduped.yml [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        tests:
+          - not_null
+
+      - name: header
+        description: A json column containing the blocks header information
```
Comment (chuxinh) [inline models/core/blocks_deduped.yml]:
```suggestion
        description: A json column containing the block's header information.
```

File: models/core/blocks_deduped.yml [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        tests:
+          - not_null
+
+      - name: ingested_at
+        description: The time this row was ingested by Chainwalkers
```
Comment (chuxinh) [inline models/core/blocks_deduped.yml]:
```suggestion
        description: The time this row was ingested by Chainwalkers.
```

File: models/core/txs_deduped.sql [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+        unique_key='tx_id',
+        tags=['core'],
+        cluster_by=['block_timestamp']
+        )
+}}
+
+with deduped_raw_txs as (
+
+    select 
+    *
```
Comment (chuxinh) [inline models/core/txs_deduped.sql]: indent

File: models/core/txs_deduped.sql [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+    from {{ source("chainwalkers", "harmony_txs" ) }}
+    where {{ incremental_load_filter("block_timestamp") }}
+    qualify row_number() over (partition by tx_id order by ingested_at desc) = 1
+
+)
+
+select * from deduped_raw_txs
```
Comment (chuxinh) [inline models/core/txs_deduped.sql]: add extra line

File: models/core/txs_deduped.yml [hunk 14]
```diff
@@ -0,0 +1,50 @@
+version: 2
+models: 
+  - name: txs_deduped
+    description: |-
+      Transactions from the Harmony chain shard 0. Duplicates have been removed.
+    columns:
+      - name: record_id
+        description: A unique id for the record generated by Chainwalkers
```
Comment (chuxinh) [inline models/core/txs_deduped.yml]:
```suggestion
        description: A unique id for the record generated by Chainwalkers.
```

File: models/core/txs_deduped.yml [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+        tests:
+          - not_null
+
+      - name: tx_id
+        description: Synonmous with transaction hash, a unique on chain identifier for the transaction
```
Comment (chuxinh) [inline models/core/txs_deduped.yml]:
```suggestion
        description: Synonmous with transaction hash, a unique on chain identifier for the transaction.
```

File: models/core/txs_deduped.yml [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+        tests:
+          - not_null
+
+      - name: tx_block_index
+        description: The index of the transaction within the block. Starts at 0.
+        tests:
+          - not_null
+
+      - name: offset_id
+        description: Synonmous with block_id for Harmony
```
Comment (chuxinh) [inline models/core/txs_deduped.yml]:
```suggestion
        description: Synonmous with block_id for Harmony.
```

File: models/core/txs_deduped.yml [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+        tests:
+          - not_null
+
+      - name: block_id
+        description: The height of the chain this block corresponds with
```
Comment (chuxinh) [inline models/core/txs_deduped.yml]:
```suggestion
        description: The height of the chain this block corresponds with.
```

File: models/core/txs_deduped.yml [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+        tests:
+          - not_null
+
+      - name: block_timestamp
+        description: The time the block was minted
```
Comment (chuxinh) [inline models/core/txs_deduped.yml]:
```suggestion
        description: The time the block was minted.
```

File: models/core/txs_deduped.yml [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+        tests:
+          - not_null
+
+      - name: network
+        description: The blockchain network (i.e. mainnet, testnet, etc.)
```
Comment (chuxinh) [inline models/core/txs_deduped.yml]:
```suggestion
        description: The blockchain network (i.e. mainnet, testnet, etc.).
```

File: models/core/txs_deduped.yml [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+        tests:
+          - not_null
+
+      - name: chain_id
+        description: Synonmous with blockchain name for Harmony
```
Comment (chuxinh) [inline models/core/txs_deduped.yml]:
```suggestion
        description: Synonmous with blockchain name for Harmony.
```

File: models/core/txs_deduped.yml [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+        tests:
+          - not_null
+         
+      - name: ingested_at
+        description: The time this row was ingested by Chainwalkers
```
Comment (chuxinh) [inline models/core/txs_deduped.yml]:
```suggestion
        description: The time this row was ingested by Chainwalkers.
```

**Review by chuxinh — APPROVED** [review]
Just some nits. Everything else look good to me.

Given now we have deduped raw tables to work with, wonder if we should adopt similar structure as what @jfmyers has suggested, similar to that of what flipside does in terms of warehouse structure.

Also give that the operation and incremental build is completed in this layer, maybe the `blocks` and even `txs` can just be a view instead. thoughts?

**Comment by drakedanner** [comment]
Before we merge this can we make sure we updated references table in existing blocks and transactions models @antonyip 

Can include it in this PR

Ping @chuxinh if you have any questions!

**Comment by antonyip** [comment]
root@9e4765c8ccfd:/harmony# dbt run
Running with dbt=0.21.1
Found 5 models, 49 tests, 0 snapshots, 0 analyses, 175 macros, 0 operations, 0 seed files, 2 sources, 0 exposures

13:16:43 | Concurrency: 4 threads (target='dev')
13:16:43 | 
13:16:43 | 1 of 5 START incremental model DEV.blocks_deduped.................... [RUN]
13:16:43 | 2 of 5 START incremental model DEV.txs_deduped....................... [RUN]
13:17:03 | 1 of 5 OK created incremental model DEV.blocks_deduped............... [SUCCESS 1 in 20.39s]
13:17:03 | 3 of 5 START incremental model DEV.blocks............................ [RUN]
13:17:24 | 3 of 5 OK created incremental model DEV.blocks....................... [SUCCESS 1 in 21.02s]
13:17:24 | 4 of 5 START table model DEV.metric_miner_count...................... [RUN]
13:17:30 | 4 of 5 OK created table model DEV.metric_miner_count................. [SUCCESS 1 in 5.91s]
13:19:11 | 2 of 5 OK created incremental model DEV.txs_deduped.................. [SUCCESS 1 in 147.66s]
13:19:11 | 5 of 5 START incremental model DEV.txs............................... [RUN]
13:19:56 | 5 of 5 OK created incremental model DEV.txs.......................... [SUCCESS 1 in 45.13s]
13:19:56 | 
13:19:56 | Finished running 4 incremental models, 1 table model in 206.68s.

Completed successfully

Done. PASS=5 WARN=0 ERROR=0 SKIP=0 TOTAL=5

@chuxinh you wanna give it one last look through?

**Pull request merged**